tiledb/sm/consolidator/fragment_consolidator.cc

```diff
@@ -407,6 +407,48 @@ Status FragmentConsolidator::consolidate_fragments(
         std::to_string(fragment_uris.size()) + " required fragments.");
   }
 
+  // In case we have a dense Array check that the fragments can be consolidated
+  // without data loss
+  if (array_for_reads->array_schema_latest().array_type() == ArrayType::DENSE) {
+    // Search every other fragment in this array if any of them overlaps in
+    // ranges and its timestamp range falls between the range of the fragments
+    // to consolidate throw error
+
+    // First calculate the timestamp bounds
+    uint64_t max_timestamp = std::numeric_limits<uint64_t>::min();
+    for (const auto& item : to_consolidate) {
+      const auto& range = item.timestamp_range();
+      max_timestamp = std::max(max_timestamp, range.second);
+    }
+
+    // Now iterate all fragments and see if the consolidation can lead to data
+    // loss
+    for (auto& frag_info : frag_info_vec) {
+      // Ignore the fragments that are requested to be consolidated
+      auto uri = frag_info.uri().last_path_part();
+      if (to_consolidate_set.count(uri) != 0) {
+        continue;
+      }
+
+      // Check domain and timestamp overlap
+      if (domain.overlap(
+              union_non_empty_domains, frag_info.non_empty_domain())) {
```

Comment:
I think we probably want to expand this to the next tiles?

Reply:
What do you mean next tiles?

Reply:
For example, if you have a single dimension array with values from 1-20 with tile extent of 5, if the computed union of non empty domains is [8, 13], the written consolidated fragment will be expanded to [6, 15] so that it includes full tiles. Here you would need to do your checks with [6, 15]. Does that make sense?

Reply:
Yes, so we need to “expand to full tiles” you are right thanks. Will update the PR later today

Reply:
Done and tested!